File: 72.edit-distance.js
/*
 * @lc app=leetcode id=72 lang=javascript
 *
 * [72] Edit Distance
 */

// @lc code=start
/**
 * @param {string} word1
 * @param {string} word2
 * @return {number}
 */
var minDistance = function (word1, word2) {
  const m = word1.length;
  const n = word2.length;
  const dp = Array.from({ length: m + 1 }, () => []);
  for (let i = 0; i <= m; ++i) {
    for (let j = 0; j <= n; ++j) {
      if (i == 0 || j == 0) {
        dp[i][j] = i === 0 ? j : i;
      } else if (word1[i - 1] == word2[j - 1]) {
        // 小心， dp[i][j] 对应word是 i-1， j-1
        dp[i][j] = dp[i - 1][j - 1];
      } else {
        dp[i][j] = Math.min(dp[i - 1][j], dp[i][j - 1], dp[i - 1][j - 1]) + 1;
      }
    }
  }
  return dp[m][n];
};
// @lc code=end

File: solutions/72.edit-distance.rb
#
# @lc app=leetcode id=72 lang=ruby
#
# [72] Edit Distance
#

# @lc code=start
# @param {String} word1
# @param {String} word2
# @return {Integer}
def min_distance(word1, word2)
  m = word1.length
  n = word2.length
  dp = []
  (0..m).each do |i|
    (0..n).each do |j|
      (dp[i] ||= [])[j] = if i.zero?
                            j
                          elsif j.zero?
                            i
                          elsif word1[i - 1] == word2[j - 1]
                            dp[i - 1][j - 1]
                          else
                            [dp[i - 1][j - 1], dp[i - 1][j], dp[i][j - 1]].min + 1
                          end
    end
  end
  dp[m][n]
end
# @lc code=end
